### PR TITLE
feature: add support for Spark Connect

### PR DESCRIPTION
**Pull request by mikita-sakalouski**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/spark/readers/test_memory.py [hunk 1]
```diff
@@ -14,10 +14,10 @@ class TestInMemoryDataReader:
         "data,format,params,expect_filter",
         [
             pytest.param(
-                "id,string\n1,hello,\n2,world", DataFormat.CSV, {"header": True}, "id < 3"
+                "id,string\n1,hello\n2,world", DataFormat.CSV, {"header":0}, "id < 3"
             ),
             pytest.param(
-                b"id,string\n1,hello,\n2,world", DataFormat.CSV, {"header": True}, "id < 3"
+                b"id,string\n1,hello\n2,world", DataFormat.CSV, {"header":0}, "id < 3"
             ),
```
Comment (dannymeijer) [inline tests/spark/readers/test_memory.py:21]: This seems like we changed our API (although perhaps in a very minor/subtle way). Let's verify if we shouldn't put backward compatibility in place here.

File: tests/spark/transformations/date_time/test_interval.py [hunk 2]
```diff
@@ -122,6 +121,8 @@ def test_interval(input_data, column_name, operation, interval, expected, spark)
 
 
 def test_interval_unhappy(spark):
+    with pytest.raises(ValueError):
+        validate_interval("some random b*llsh*t")  # TODO: this should raise an error, but it doesn't in REMOTE mode
```
Comment (dannymeijer) [inline tests/spark/transformations/date_time/test_interval.py:125]: I think we can remove the comment here

File: src/koheesio/models/sql.py [hunk 3]
```diff
@@ -28,8 +29,8 @@ class SqlBaseStep(Step, ExtraParamsMixin, ABC):
     sql: Optional[str] = Field(default=None, description="SQL script to apply")
     params: Dict[str, Any] = Field(
         default_factory=dict,
-        description="Placeholders (parameters) for templating. These are identified with ${placeholder} in the SQL "
-        "script. Note: any arbitrary kwargs passed to the class will be added to params.",
+        description="Placeholders (parameters) for templating. The substitutions are identified by braces ('{' and '}')"
+        "and can optionally contain a $-sign. Note: any arbitrary kwargs passed to the class will be added to params.",
```
Comment (dannymeijer) [inline src/koheesio/models/sql.py:33]: I'm not sure if the current implementation matches the description (as we uncommented the replace params part)

**Review by dannymeijer — DISMISSED** [review]

File: src/koheesio/spark/delta.py [hunk 4]
```diff
@@ -300,7 +299,9 @@ def exists(self) -> bool:
         result = False
 
         try:
-            self.spark.table(self.table_name)
+            # In Spark remote session it is not enough to call just spark.table(self.table_name)
+            # as it will not raise an exception, we have to make action call on table to check if it exists
+            self.spark.table(self.table_name).take(1)
```
Comment (pariksheet) [inline src/koheesio/spark/delta.py:304]: can we run self.spark.table(self.table_name).take(1) only for remote session.

This is expensive operation for just to check if table exists or not

Comment (mikita-sakalouski) [reply src/koheesio/spark/delta.py:304]: Fixed

File: src/koheesio/models/reader.py [hunk 5]
```diff
@@ -2,13 +2,12 @@
...
-from typing import Optional, TypeVar
 from abc import ABC, abstractmethod
+from typing import Optional
 
-from koheesio import Step
 
-# Define a type variable that can be any type of DataFrame
-DataFrameType = TypeVar("DataFrameType")
+from koheesio import Step
+from koheesio.spark import DataFrame
```
[the remark this answers is not in the document]
Comment (mikita-sakalouski) [reply src/koheesio/models/reader.py:9]: Rollback

File: src/koheesio/spark/__init__.py [hunk 6]
```diff
@@ -56,11 +60,7 @@ def _get_active_spark_session(self):
         attempted to be retrieved.
         """
         if self.spark is None:
-            self.spark = SparkSession.getActiveSession()
-        return self
+            from koheesio.spark.utils.common import get_active_session
 
-
-# TODO: Move to spark/functions/__init__.py after reorganizing the code
```
[the remark this answers is not in the document]
Comment (mikita-sakalouski) [reply src/koheesio/spark/__init__.py:63]: Put back reference

File: src/koheesio/spark/transformations/sql_transform.py [hunk 7]
```diff
@@ -30,8 +31,14 @@ def execute(self):
         table_name = get_random_string(prefix="sql_transform")
         self.params = {**self.params, "table_name": table_name}
 
-        df = self.df
-        df.createOrReplaceTempView(table_name)
+        from koheesio.spark.utils.connect import is_remote_session
+
+        if 3.4 < SPARK_MINOR_VERSION < 4.0 and is_remote_session() and self.df.isStreaming:
+            raise RuntimeError("""SQL Transform is not supported in remote sessions with streaming dataframes.
```
[the remark this answers is not in the document]
Comment (mikita-sakalouski) [reply src/koheesio/spark/transformations/sql_transform.py]: Done

File: src/koheesio/spark/writers/delta/stream.py [hunk 8]
```diff
@@ -32,5 +32,7 @@ class Options(BaseModel):
     def execute(self):
         if self.batch_function:
             self.streaming_query = self.writer.start()
+        # elif self.streaming and self.is_remote_spark_session:
+        #     self.streaming_query = self.writer.start()
```
[the remark this answers is not in the document]
Comment (mikita-sakalouski) [reply src/koheesio/spark/writers/delta/stream.py:36]: Removed

File: tests/spark/readers/test_memory.py (hunk 1 above)
Comment (mikita-sakalouski) [reply tests/spark/readers/test_memory.py:21]: Fixed, put some code for backward compatibility

File: tests/spark/transformations/date_time/test_interval.py (hunk 2 above)
Comment (mikita-sakalouski) [reply tests/spark/transformations/date_time/test_interval.py:126]: Fixed

File: src/koheesio/models/sql.py (hunk 3 above)
Comment (mikita-sakalouski) [reply src/koheesio/models/sql.py:33]: We still have: `query = query.replace(f"${{{key}}}", value)` .  So looks like we are good from comment side

**Review by dannymeijer — APPROVED** [review]

**Pull request merged**